lib/request.js

```diff
@@ -23,6 +24,100 @@ function refreshAuth(type, config) {
   });
 }
 
+const execChannels = [
+  'stdin',
+  'stdout',
+  'stderr',
+  'error',
+  'resize'
+];
+
+/**
+ * Determine whether a failed Kubernetes API response is asking for an upgrade
+ * @param {object} body - response body object from Kubernetes
+ * @property {string} status - request status
+ * @property {number} code - previous request's response code
+ * @property {message} message - previous request response message
+ * @returns {boolean} Upgrade the request
+ */
+
+function isUpgradeRequired(body) {
+  return body.status === 'Failure'
+    && body.code === 400
+    && body.message === 'Upgrade request required';
+}
+
+/**
+ * Construct the query parameters necessary for performing an execAction
+ * @param {object} qs - query parameter object as passed by user
+ * @property {string[]} command - (required) command and arguments to be executed
+ * @property {string} container - (optional) container in which to execute the command; optional if there's only one container in the pod
+ * @property {boolean} tty - (optional) whether to hold the websocket connection open for an interactive session (not yet implemented)
+ * @param {function} cb - callback to handle the response
+ * @returns {string} All necessary query parameters as a string ready for inclusion in URL
+ */
+function buildExecQueryParams(qs, cb) {
+  const command = qs.command;
+  if (!command || !Array.isArray(command)) {
+    return cb(new Error(`expected command to be an array of strings, not: ${JSON.stringify(command)}`));
+  }
+
+  // The command query parameter gets specified multiple times to
+  // indicate arguments to the initial binary
+  // (command=ls&command=-al), so it cannot be handled in an object
+  // like the rest of the qs.
+  const commandAsParams = command.map((cmd) => `command=${encodeURIComponent(cmd)}`);
+  delete qs.command;
```

Comment:
I'd prefer not to mutate here if possible, but I assume something like 

```js
const { qs: { command, ...rest } } = options;
```

is out of the question? :P

Reply:
I think the `qs` suggestion about might help resolve this.